### DISCUSSION TITLE
用Python玩Premutation

### DISCUSSION THREAD

**Discussion by mrweixd** [1101-GS4719-C/Python-Community-GS4719-C#1603] [General]
還記得高中的排列組合嗎？解不出來時最好的朋友就是用樹狀圖暴力解。
只要寫出每種可能性就穩妥LA。
廢話不多說，先上程式碼：
```
def perm(  pers,  i , n ):
    if(i==n):
        for j in range(n):
            print(pers[j],end='')
        print("  ")
    else:
        for j in range(i,n):
            pers[i],pers[j]=pers[j],pers[i]
            perm(pers, i+1, n)
            pers[i],pers[j]=pers[j],pers[i]
            

pers= []
temp=input('Input pers:')
for i in range(len(temp)):
    pers.append(temp[i])
    
perm(pers,0,len(temp))
```
先上實際執行畫面：
![image](https://github.com/NCU-GS4719-Python/Python-Community/assets/76525995/92117b5e-8f99-4134-bd3c-8503cf939ac8)
先輸入要排列的字串，範例輸入abc：
![image](https://github.com/NCU-GS4719-Python/Python-Community/assets/76525995/b009a66a-f6ee-42a0-96d0-017da0236a92)
Enter後就會印出所有排列了！
![image](https://github.com/NCU-GS4719-Python/Python-Community/assets/76525995/b62b635e-5490-4eb2-994e-ce86c028184f)

**Comment by v0ouo** [comment]
用遞迴也是一種很好的方法，這裡推薦一個python工具itertools.permutations()，可以直接列出所有可能性
程式碼:
```import itertools
pers= []
temp=input('Input pers:')
for i in range(len(temp)):
    pers.append(temp[i])
for i in itertools.permutations(pers, len(temp)):
    print(i)
```

輸出:
![螢幕擷取畫面 2023-09-27 151431](https://github.com/NCU-GS4719-Python/Python-Community/assets/144879969/27aea255-8f28-4bc8-b076-0f98e6e7c8e9)

**Comment by poyen0806** [comment]
對不起，但我還是想說一下標題
是permutation而不是permutation

另外我覺得這個方法不錯
很適合回憶痛苦的排列組合
還好我不是在修機率與統計的時候滑到這則貼文
不然我會瘋掉

遞迴的方法也不錯
那你們知道在Google查遞迴的話會出現什麼嗎
沒錯
他會問你是不是要查遞迴
![image](https://github.com/user-attachments/assets/aae5d652-65bb-49ff-98e1-1b708996eab6)
